cannot load commit.template

Thanks for the great product!
I just started using it, but I'm reporting something I didn't quite understand.

I have commit.template set up and it is loaded when I run the `git commit` command from the shell.
However, when I try to edit the commit message by `^e` when committing from gitui, it is not loaded.

I am aware that commit.template is supported, but is there something wrong with my configuration? (https://github.com/extrawurst/gitui/issues/546)

The first one is when you run `git commit`, and the second one is when you open the editor from gitui.

<img width="527" alt="image" src="https://github.com/extrawurst/gitui/assets/38337195/8542a4f1-25a2-41fb-b0c9-10f84084143f">
<img width="396" alt="image" src="https://github.com/extrawurst/gitui/assets/38337195/af7ba1c0-ff4c-42df-b75a-be8013838b03">

The following is a description of the settings for git commit.template.

<img width="307" alt="image" src="https://github.com/extrawurst/gitui/assets/38337195/958942a0-4097-4901-823a-de90dc0f0b74">
<img width="294" alt="image" src="https://github.com/extrawurst/gitui/assets/38337195/d4f09e42-6cde-4722-ab5a-f979cb9bee8d">

- OS: macOS Ventura 13.3
- GitUI version: 0.23.0

@extrawurst 

Totally missed that. Thanks!
Here's the error part from the generated log.

```
06:48:18 [TRACE] (1) gitui::app: [src/app.rs:509] update_async: Git(Status)
06:48:18 [TRACE] (1) gitui::app: [src/app.rs:392] event: Input(Key(KeyEvent { code: Char('c'), modifiers: NONE, kind: Press, state: NONE }))
06:48:18 [ERROR] read commit.template failed: No such file or directory (os error 2)
06:48:21 [TRACE] (1) gitui::app: [src/app.rs:392] event: Input(Key(KeyEvent { code: Char('e'), modifiers: CONTROL, kind: Press, state: NONE }))
06:48:22 [INFO] input polling suspended
06:48:22 [TRACE] (1) gitui::app: [src/app.rs:392] event: State(Paused)
06:48:25 [INFO] input polling resumed
```

The result showed an error indicating that it couldn't detect the presence of the file.

Is there any other information you need from me?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -24,6 +24,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * fix commit dialog char count for multibyte characters ([#1726](https://github.com/extrawurst/gitui/issues/1726))
 * fix wrong hit highlighting in fuzzy find popup [[@UUGTech](https://github.com/UUGTech)] ([#1731](https://github.com/extrawurst/gitui/pull/1731))
 * fix symlink support for configuration files [[@TheBlackSheep3](https://github.com/TheBlackSheep3)] ([#1751](https://github.com/extrawurst/gitui/issues/1751))
+* expand `~` in `commit.template` ([#1745](https://github.com/extrawurst/gitui/pull/1745))
 
 ## [0.23.0] - 2022-06-19
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -46,6 +46,7 @@ ron = "0.8"
 scopeguard = "1.2"
 scopetime = { path = "./scopetime", version = "0.1" }
 serde = "1.0"
+shellexpand = "3.1"
 simplelog = { version = "0.12", default-features = false }
 struct-patch = "0.2"
 syntect = { version = "5.0", default-features = false, features = ["parsing", "default-syntaxes", "default-themes", "html"] }

--- a/src/components/commit.rs
+++ b/src/components/commit.rs
@@ -30,6 +30,8 @@ use ratatui::{
 use std::{
 	fs::{read_to_string, File},
 	io::{Read, Write},
+	path::PathBuf,
+	str::FromStr,
 };
 
 enum CommitResult {
@@ -364,61 +366,80 @@ impl CommitComponent {
 
 		let repo_state = sync::repo_state(&self.repo.borrow())?;
 
-		self.mode =
-			if repo_state != RepoState::Clean && reword.is_some() {
-				bail!("cannot reword while repo is not in a clean state");
-			} else if let Some(reword_id) = reword {
-				self.input.set_text(
-					sync::get_commit_details(
+		self.mode = if repo_state != RepoState::Clean
+			&& reword.is_some()
+		{
+			bail!("cannot reword while repo is not in a clean state");
+		} else if let Some(reword_id) = reword {
+			self.input.set_text(
+				sync::get_commit_details(
+					&self.repo.borrow(),
+					reword_id,
+				)?
+				.message
+				.unwrap_or_default()
+				.combine(),
+			);
+			self.input.set_title(strings::commit_reword_title());
+			Mode::Reword(reword_id)
+		} else {
+			match repo_state {
+				RepoState::Merge => {
+					let ids =
+						sync::mergehead_ids(&self.repo.borrow())?;
+					self.input
+						.set_title(strings::commit_title_merge());
+					self.input.set_text(sync::merge_msg(
 						&self.repo.borrow(),
-						reword_id,
-					)?
-					.message
-					.unwrap_or_default()
-					.combine(),
-				);
-				self.input.set_title(strings::commit_reword_title());
-				Mode::Reword(reword_id)
-			} else {
-				match repo_state {
-					RepoState::Merge => {
-						let ids =
-							sync::mergehead_ids(&self.repo.borrow())?;
-						self.input
-							.set_title(strings::commit_title_merge());
-						self.input.set_text(sync::merge_msg(
-							&self.repo.borrow(),
-						)?);
-						Mode::Merge(ids)
-					}
-					RepoState::Revert => {
-						self.input
-							.set_title(strings::commit_title_revert());
-						self.input.set_text(sync::merge_msg(
-							&self.repo.borrow(),
-						)?);
-						Mode::Revert
-					}
-
-					_ => {
-						self.commit_template = get_config_string(
-							&self.repo.borrow(),
-							"commit.template",
-						)
-						.ok()
-						.flatten()
-						.and_then(|path| read_to_string(path).ok());
-
-						if self.is_empty() {
-							if let Some(s) = &self.commit_template {
-								self.input.set_text(s.clone());
-							}
-						}
-						self.input.set_title(strings::commit_title());
-						Mode::Normal
-					}
+					)?);
+					Mode::Merge(ids)
 				}
-			};
+				RepoState::Revert => {
+					self.input
+						.set_title(strings::commit_title_revert());
+					self.input.set_text(sync::merge_msg(
+						&self.repo.borrow(),
+					)?);
+					Mode::Revert
+				}
+
+				_ => {
+					self.commit_template = get_config_string(
+						&self.repo.borrow(),
+						"commit.template",
+					)
+					.map_err(|e| {
+						log::error!("load git-config failed: {}", e);
+						e
+					})
+					.ok()
+					.flatten()
+					.and_then(|path| {
+						shellexpand::full(path.as_str())
+							.ok()
+							.and_then(|path| {
+								PathBuf::from_str(path.as_ref()).ok()
+							})
+					})
+					.and_then(|path| {
+						read_to_string(&path)
+							.map_err(|e| {
+								log::error!("read commit.template failed: {e} (path: '{:?}')",path);
+								e
+							})
+							.ok()
+					});
+
+					if self.is_empty() {
+						if let Some(s) = &self.commit_template {
+							self.input.set_text(s.clone());
+						}
+					}
+					self.input.set_title(strings::commit_title());
+					Mode::Normal
+				}
+			}
+		};
 
 		self.commit_msg_history_idx = 0;
 		self.input.show()?;

